nomos-services/cryptarchia-consensus/src/lib.rs

```diff
@@ -872,6 +854,206 @@ where
             blobs = blobs
         );
     }
+
+    /// Retrieves the blocks in the range from `from` to `to` from the storage.
+    /// Both `from` and `to` are included in the range.
+    /// This is implemented here, and not as a method of `StorageAdapter`, to simplify the panic
+    /// and error message handling.
+    ///
+    /// # Panics
+    ///
+    /// Panics if any of the blocks in the range are not found in the storage.
+    ///
+    /// # Parameters
+    ///
+    /// * `from` - The header id of the first block in the range. Must be a valid header.
+    /// * `to` - The header id of the last block in the range. Must be a valid header.
+    ///
+    /// # Returns
+    ///
+    /// A vector of blocks in the range from `from` to `to`.
+    /// If no blocks are found, returns an empty vector.
+    /// If any of the HeaderId are invalid, returns an error with the first invalid header id.
+    async fn get_blocks_in_range(
+        from: HeaderId,
+        to: HeaderId,
+        storage_adapter: &StorageAdapter<Storage, TxS::Tx, BS::BlobId>,
+    ) -> Vec<Block<ClPool::Item, DaPool::Item>> {
```

Comment:
Could this also be an iterator and letting the consumers choose?

Reply:
Dani asked me the same thing. Because I answered there, I'd rather have the convo all in the same place if you don't mind :)
https://github.com/logos-co/nomos-node/pull/1043#discussion_r1975951506